File: BACK/Nest/src/external/external.controller.ts
import { Controller, Get, NotFoundException, Param, ParseIntPipe, Query } from '@nestjs/common';
import { ApiBearerAuth, ApiOperation, ApiResponse, ApiTags } from '@nestjs/swagger';
import { Public } from '../common/decorators/public.decorator';
import { ExternalPost, PostWithComments } from './dto/external-post.dto';
import { ExternalService } from './external.service';

/**
 * Exemple de routes qui appellent une API publique tierce (JSONPlaceholder,
 * https://jsonplaceholder.typicode.com) via ExternalService (HttpModule
 * typé + retries). Montre comment agréger/relayer des données externes
 * depuis notre propre API.
 */
@ApiTags('external')
@Controller('external')
export class ExternalController {
  constructor(private readonly externalService: ExternalService) {}

  /** Liste des posts de l'API publique, avec filtre optionnel par userId. */
  @Public()
  @Get('posts')
  @ApiOperation({ summary: 'Liste des posts (proxy JSONPlaceholder)' })
  getPosts(@Query('userId') userId?: string): Promise<ExternalPost[]> {
    const parsedUserId = userId !== undefined ? Number(userId) : undefined;
    return this.externalService.getPosts(parsedUserId);
  }

  /** Detail d'un post de l'API publique. */
  @Public()
  @Get('posts/:id')
  @ApiOperation({ summary: "Detail d'un post (proxy JSONPlaceholder)" })
  @ApiResponse({ status: 404, description: 'Non trouve' })
  async getPost(@Param('id', ParseIntPipe) id: number): Promise<ExternalPost> {
    const post = await this.externalService.getPost(id);
    if (!post) {
      throw new NotFoundException();
    }
    return post;
  }

  /**
   * Route protegee qui combine deux appels externes (post + commentaires).
   * Necessite un token JWT : montre comment mixer authentification et appels sortants.
   */
  @Get('posts/:id/with-comments')
  @ApiBearerAuth()
  @ApiOperation({ summary: 'Post + commentaires (proxy JSONPlaceholder, protege)' })
  @ApiResponse({ status: 401, description: 'Non authentifie' })
  @ApiResponse({ status: 404, description: 'Non trouve' })
  async getPostWithComments(@Param('id', ParseIntPipe) id: number): Promise<PostWithComments> {
    const post = await this.externalService.getPost(id);
    if (!post) {
      throw new NotFoundException();
    }

    const comments = await this.externalService.getCommentsForPost(id);
    return { post, comments };
  }
}

File: BACK/Dotnet/src/ApiRest.Api/Controllers/ExternalController.cs
using ApiRest.Api.Models.External;
using ApiRest.Api.Services.External;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace ApiRest.Api.Controllers;

/// <summary>
/// Exemple de routes qui font appel à une API publique tierce (JSONPlaceholder,
/// https://jsonplaceholder.typicode.com) via un HttpClient typé (voir
/// Services/External/JsonPlaceholderClient.cs). Montre comment agréger/relayer des
/// données externes depuis notre propre API.
/// </summary>
[ApiController]
[Route("api/external")]
public class ExternalController : ControllerBase
{
    private readonly IJsonPlaceholderClient _client;

    public ExternalController(IJsonPlaceholderClient client)
    {
        _client = client;
    }

    /// <summary>Liste des posts de l'API publique, avec filtre optionnel par userId.</summary>
    [HttpGet("posts")]
    [AllowAnonymous]
    [ProducesResponseType(typeof(IReadOnlyList<ExternalPost>), StatusCodes.Status200OK)]
    public async Task<ActionResult<IReadOnlyList<ExternalPost>>> GetPosts(
        [FromQuery] int? userId, CancellationToken cancellationToken)
    {
        var posts = await _client.GetPostsAsync(userId, cancellationToken);
        return Ok(posts);
    }

    /// <summary>Détail d'un post de l'API publique.</summary>
    [HttpGet("posts/{id:int}")]
    [AllowAnonymous]
    [ProducesResponseType(typeof(ExternalPost), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<ActionResult<ExternalPost>> GetPost(int id, CancellationToken cancellationToken)
    {
        var post = await _client.GetPostAsync(id, cancellationToken);
        return post is null ? NotFound() : Ok(post);
    }

    /// <summary>
    /// Route protégée qui combine deux appels externes (post + commentaires).
    /// Nécessite un token JWT : montre comment mixer authentification et appels sortants.
    /// </summary>
    [HttpGet("posts/{id:int}/with-comments")]
    [Authorize]
    [ProducesResponseType(typeof(PostWithComments), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<ActionResult<PostWithComments>> GetPostWithComments(int id, CancellationToken cancellationToken)
    {
        var post = await _client.GetPostAsync(id, cancellationToken);
        if (post is null)
        {
            return NotFound();
        }

        var comments = await _client.GetCommentsForPostAsync(id, cancellationToken);
        return Ok(new PostWithComments(post, comments));
    }
}
